core/src/main/java/com/graphhopper/reader/dem/EdgeElevationSmoothing.java

```diff
@@ -4,58 +4,115 @@
 import com.graphhopper.util.DistancePlaneProjection;
 import com.graphhopper.util.PointList;
 
+import java.util.HashMap;
+
 /**
- * The ElevationData is read from rectangular tiles. Especially when going along a cliff,
- * valley, or pass, it can happen that a small part of the road contains incorrect elevation data.
- * This is because the elevation data is coarse and sometimes contains errors.
+ * Elevation data is read from DEM tiles that have data points for rectangular tiles usually having an
+ * edge length of 30 or 90 meter. Elevation in between the middle points of those tiles will be
+ * interpolated and weighted by the distance from a node to adjacent tile centers.
+ * <p>
+ * Ways that go along cliffs or ridges are particularly affected by ups and downs that do not reflect
+ * the actual elevation but may be artifacts originated from very accurately mapping when elevation has
+ * a lower resolution.
  *
- * @author Robin Boldt
+ * @author Peter Karich
+ * @author Christoph Lingg
  */
 public class EdgeElevationSmoothing {
+    public static void smoothMovingAverage(PointList geometry, double maxWindowSize) {
+        // calculate the distance between all points once here to avoid repeated calculation.
+        // for n nodes there are always n-1 edges
+        double[] distances = new double[geometry.size() - 1];
+        for (int i = 0; i <= geometry.size() - 2; i++) {
+            distances[i] = DistancePlaneProjection.DIST_PLANE.calcDist(
+                    geometry.getLat(i), geometry.getLon(i),
+                    geometry.getLat(i + 1), geometry.getLon(i + 1)
+            );
+        }
 
-    // If the point is farther then this, we stop averaging
-    private final static int MAX_SEARCH_DISTANCE = 150;
 
-    /**
-     * This method smooths the elevation data of a PointList by calculating the average elevation over
-     * multiple points of that PointList.
-     */
-    public static void smoothMovingAverage(PointList geometry) {
-        for (int i = 1; i < geometry.size() - 1; i++) {
+        // map that will collect all smoothed elevation values, size is less by 2
+        // because elevation of start and end point (tower nodes) won't be touched
+        HashMap<Integer, Double> averagedElevations = new HashMap<>(geometry.size() - 2);
```

Comment:
Or you could simply use a copy of the initial elevation data and avoid map semantics entirely?

```java
double[] averagedElevations = new double[geometry.size()];
for (int i = 0; i < geometry.size(); i++) {
   averagedElevations[i] = geometry.getEle(i);
}
```

Reply:
good point about the rehashing. I also tried out avoiding a map but I was also not happy because you 1) copy a lot of values that are later on overwritten and 2) I don't like that we overwrite (with the exact same value) the evaluation of tower nodes which conceptually seems wrong.

all that said, i have no strong feelings here

Reply:
regarding performance: while more complex, this algorithm is faster than the previous and simpler algorithm which suffered from recalculations

Reply:
Could you still switch to a `IntDoubleHashMap` ? The regular HashMap adds a lot of pointless Auto(un)boxing.

Reply:
I just pushed a change, actually I forgot to include that change last Friday.